### DISCUSSION TITLE
Benefits of transferring data to GPU before passing it to `NeighborLoader`

### DISCUSSION THREAD

**Discussion by liadgerstman** [pyg-team/pytorch_geometric#7847] [Q&A] [answered]
Hi, I understand that neighbor sampling using NeighborLoader is supposed to mainly  run on CPU and then we need to transfer the sampled batches from the CPU to the GPU for the training part. 
So: 
1. Why is it claimed in reddit.py that the nodes' features/labels are sent to the GPU 'for faster access during sampling'? Does it mean that the sampling is being done on the GPU in this case? I am asking this because I could not see the GPU kernels doing sampling when profiling this case.
![image](https://github.com/pyg-team/pytorch_geometric/assets/61936005/effdf18e-683c-4f18-a15e-f1e8ba3c9f69). If the sampling is indeed being processed on a CPU- why aren't the batches being sent to the GPU as well?
 
2. I wanted to clarify that the sampling process itself only starts when the next() is called when we iterate over NeighborLoader iterator and not we construct NeighborLoader. Hence, the sampling is not done here:
![image](https://github.com/pyg-team/pytorch_geometric/assets/61936005/18b9aecc-0179-4769-9d90-b1565ad0ee0e)
but here: 
![image](https://github.com/pyg-team/pytorch_geometric/assets/61936005/33055742-f9a3-4ff1-b0d5-3c659009ca4b)

Thank you for the clarification

**Comment by akihironitta** [comment]
Hey @liadgerstman! I haven't benchmarked this, but I guess there should be some speedup. AFAIK, this `NeighborLoader` (1) samples node indices first on CPU, and (2) packs node features and labels (`data.{x,y}`) of sampled node indices into cuda tensors directly instead of CPU tensors. If you don't transfer the `data.{x,y}` to the device beforehand, you will have to transfer the sampled data `next(train_loader_iter).{x,y}` to the device in each iteration, which may create performance difference.

**Comment by liadgerstman** [comment]
Hey @akihironitta, I profiled reddit.py and it looks like the gpu is idle during NeighborLoader execution - only CPU is involved (nodes' features/labels are sent to the gpu before it starts). That is why I asked the questions:
Is the sampling process actually being executed when NeighborLoader is called here? Or is it done lazily by the time the next() function is called while you iterate over the train_loader in the for loop.
![image](https://github.com/pyg-team/pytorch_geometric/assets/61936005/c6e88961-aef5-4471-9b7c-8ed6074f228d) In either ways, the gpu is not involved in the sampling process (according to the profiling) and that makes the comment in the reddit.py quiet confusing (The comment that says that nodes' features/labels are sent to the GPU 'for faster access during sampling). 
I am interested also in understanding if gpu sampling is even possible using PYG's NeighborLoader- I tried to pass dataset[0] to the gpu before calling to NeighborLoader but it fails when you call NeighborLoader. When you only transfer x, y, edge_index to the gpu it doesn't fail but it looks like the sampling is still done on the cpu.

**Reply by akihironitta** [reply]
Hey @liadgerstman, thanks for sharing your questions on this! :)

> it looks like the gpu is idle during NeighborLoader execution

Yes, neighbour sampling itself doesn't happen on GPU.

> Is the sampling process actually being executed when NeighborLoader is called here?

No.

> Or is it done lazily by the time the next() function is called while you iterate over the train_loader in the for loop.

Yes, the sampling procedure is triggered when you call `next(loader_iter)` in each iteration.

> I am interested also in understanding if gpu sampling is even possible using PYG's NeighborLoader

AFAIK, we don't have neighbour sampling supported on GPU yet:
- tracking issue: https://github.com/pyg-team/pyg-lib/issues/23
- source code: https://github.com/pyg-team/pyg-lib/tree/6e9e3b8bf58ce4e73a39b3ed0673541f7ad69d90/pyg_lib/csrc/sampler

> that makes the comment in the reddit.py quiet confusing (The comment that says that nodes' features/labels are sent to the GPU 'for faster access during sampling).

How do you think we should update the comment? Would this be clearer?
```python
# features/labels are sent to GPU so you don't have to in each iteration
```

---

What I meant here https://github.com/pyg-team/pytorch_geometric/discussions/7847#discussioncomment-6635919 is that if you don't transfer your `data.{x,y}` in advance, you'll have to move each `sampled_data.{x,y}` in each iteration, which can slow down your training.

Just for your information, I've profiled a similar script, and I see a bunch of slow host-to-device transfers (30-40ms for each transfer).

**1. With `data` transferred to GPU before iterations** (70-90ms per iter)
<img width="2671" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/20610905/12841127-d9c8-456a-9866-414d4a5af834">

**2. With `data` transferred to GPU in each iteration** (140-172ms per iter)
<img width="2748" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/20610905/9ba0a36a-c46e-4e06-a959-95d735cf4af9">

<details><summary>code</summary>

```python
import os.path as osp
import datetime

import torch
from torch_geometric.datasets import Reddit
from torch_geometric.loader import NeighborLoader


def main1():
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'Reddit')
    dataset = Reddit(path)
    data = dataset[0].to(device, 'x', 'y')
    loader = NeighborLoader(
        data,
        input_nodes=data.train_mask,
        num_neighbors=[25, 10],
        shuffle=True,
        batch_size=1024,
    )
    loader_iter = iter(loader)

    with torch.profiler.profile(with_stack=True) as p:
        for _ in range(5):
            batch = next(loader_iter)
            batch.to(device, 'x', 'y')  # no-op as data is already on device

    now = datetime.datetime.now().strftime("%Y_%m_%d:%H.%M.%S.pt.trace.json.gz")
    p.export_chrome_trace(f"main1_{now}")


def main2():
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'Reddit')
    dataset = Reddit(path)
    data = dataset[0]
    loader = NeighborLoader(
        data,
        input_nodes=data.train_mask,
        num_neighbors=[25, 10],
        shuffle=True,
        batch_size=1024,
    )
    loader_iter = iter(loader)

    with torch.profiler.profile(with_stack=True) as p:
        for _ in range(5):
            batch = next(loader_iter)
            batch.to(device, 'x', 'y')

    now = datetime.datetime.now().strftime("%Y_%m_%d:%H.%M.%S.pt.trace.json.gz")
    p.export_chrome_trace(f"main2_{now}")


if __name__ == '__main__':
    main1()
    main2()
```

</details>

**Reply by rusty1s** [reply]
Yes, exactly. The major speed improvements are:
* `index_select` for feature fetching is faster on GPU than on CPU
* We avoid the unnecessary device transfer

**Reply by liadgerstman** [reply]
Hi @akihironitta  and @rusty1s thanks for your answers!
So, to sum things up,

- The comment : # features/labels are sent to GPU so you don't have to in each iteration is now much better 👍
- Regarding @akihironitta 's profiling (thank you for looking at that!), we can see that data transfers being executed in between the iterations slow down the whole training process, and we can see that nothing is being executed on the GPU except for some overhead functionality probably. By the way, why not transferring the edge_index to the GPU as well before sampling? (I guess that it is because we want to avoid device transfer of unnecessary edge_index elements?)
- The neighbor sampling is being done lazily when next() is called.

Thank you again for the great answers

**Reply by akihironitta** [reply]
> The comment : # features/labels are sent to GPU so you don't have to in each iteration is now much better 👍

I missed this until @rusty1s commented above, but the current comment is still true (and clear enough IMO) as `index_select` runs faster on GPU as you can see in the profiling result. Do you have any suggestion on how to update the comment? Feel free to send a PR if any :)

> why not transferring the edge_index to the GPU as well before sampling?

I think there won't be much difference because sampling node indices, which uses `edge_index`, will run on CPU anyway as sampling on GPU isn't supported yet (https://github.com/pyg-team/pyg-lib/issues/23).

**Reply by liadgerstman** [reply]
@akihironitta, do you suggest that the comment is still true because the index_select is being done on the GPU? Is that the only thing being done on the GPU in the context of sampling? 
I was thinking that since the sampling itself (except for the final index_select) is currently not executed on the GPU , it is a bit confusing saying that transferring the data to the GPU is because of 'fast access during sampling'. I would suggest giving the context of index_select in the comment. BTW, now that we have this discussion here I guess it is no longer an issue.

**Reply by akihironitta** [reply, marked as answer]
> do you suggest that the comment is still true because the index_select is being done on the GPU? Is that the only thing being done on the GPU in the context of sampling?

Yes, and Yes from what I see from the profiler. You can also check it on your own by running my code snippet above.

Thanks for sharing your thoughts. Feel free to send a PR if you have any suggestions around the comment :)
